mesonbuild/cmake/executor.py

```diff
@@ -130,39 +134,105 @@ def check_cmake(self, cmakebin: 'ExternalProgram') -> T.Optional[str]:
         cmvers = re.sub(r'\s*cmake version\s*', '', out.split('\n')[0]).strip()
         return cmvers
 
+    def set_exec_mode(self, print_cmout: T.Optional[bool] = None, always_capture_stderr: T.Optional[bool] = None) -> None:
+        if print_cmout is not None:
+            self.print_cmout = print_cmout
+        if always_capture_stderr is not None:
+            self.always_capture_stderr = always_capture_stderr
+
     def _cache_key(self, args: T.List[str], build_dir: str, env):
         fenv = frozenset(env.items()) if env is not None else None
         targs = tuple(args)
         return (self.cmakebin, targs, build_dir, fenv)
 
-    def _call_real(self, args: T.List[str], build_dir: str, env) -> T.Tuple[int, str, str]:
+    def _call_cmout_stderr(self, args: T.List[str], build_dir: str, env) -> TYPE_result:
+        cmd = self.cmakebin.get_command() + args
+        proc = S.Popen(cmd, stdout=S.PIPE, stderr=S.PIPE, cwd=build_dir, env=env)
+
+        # stdout and stderr MUST be read at the same time to avoid pipe
+        # blocking issues. The easiest way to do this is with a separate
+        # thread for one of the pipes.
+        def print_stdout():
+            while True:
+                line = proc.stdout.readline()
+                if not line:
+                    break
+                mlog.log(line.decode('utf-8').strip('\n'))
+            proc.stdout.close()
+
+        t = Thread(target=print_stdout)
+        t.start()
```

Comment:
If anything in the below block throws, the `join` is never run. The joining should probably be done in a `finally` block.

Similarly `line.decode('utf-8')` throws on non-utf8 input. Is CMake guaranteed to always print UTF-8 only?

Reply:
Fixed. Thanks.

I doubt that CMake produces non-UTF-8 output, but I decided to change it to `errors='ignore'` just in case.